### PR TITLE
Support datastreams as an AuditLog Sink

### PR DESCRIPTION
**Pull request by tmanninger**
### Description

This PR allows writing auditlogs to datastreams.

 auditlog type "internal_opensearch" currently not supporting datatstreams. 
`java.lang.IllegalArgumentException: only write ops with an op_type of create are allowed in data streams
`

This PR changes the op_type to "create" for the "internal_opensearch" auditlogs, because audit logs should never be updated.
Datastreams only supporting "create" as op_type.

### Issues Resolved
- https://github.com/opensearch-project/security/issues/3745

### Testing
Tested manually: Security auditlog with and without datastreams

### CODE REVIEW

**Comment by peternied** [comment]
@tmanninger Thanks for the contribution,  can you please since your commits with the DCO [1]?

- [1] https://github.com/opensearch-project/security/pull/4257/checks?check_run_id=23865229124

File: src/main/java/org/opensearch/security/auditlog/sink/InternalOpenSearchSink.java
Comment (peternied) [inline src/main/java/org/opensearch/security/auditlog/sink/InternalOpenSearchSink.java:89]: Please add new suite of test cases for audit logs that uses datastream.  I would recommend extending the existing 'basic' test cases [1] and modify them slightly to set a datastream instead of the default index.

- [1] https://github.com/opensearch-project/security/blob/main/src/test/java/org/opensearch/security/auditlog/integration/BasicAuditlogTest.java

**Review by peternied — CHANGES_REQUESTED** [review]

**Review by DarshitChanpura — COMMENTED** [review]
+1 to Peter's comments. @tmanninger Could you please +ve/-ve test cases for this change. We would like to really ensure that explicitly marking op_type as CREATE would resolve the question.

**Comment by willyborankin** [comment]
Agree with @peternied and @DarshitChanpura . The data stream can be created via `client.createDataStream`. The setting `OpType.CREATED` for the document can help but currently an audit index is not a data stream. 
To create a data stream first you need:
 - Create an index template since data stream index is different compare to the regular index
 - Create a data stream
 - Create ISM polity to rollover and clean data stream indices
 
Here is an examle for it:
 https://github.com/opensearch-project/opensearch-java/blob/main/samples/src/main/java/org/opensearch/client/samples/DataStreamBasics.java

**Comment by willyborankin** [comment]
I think the right solution would be to extend existing `InternalOpenSearchSink` or crate a new `Sync` call it let's say `internal_opensearch_datastream` or something else.  @peternied, @DarshitChanpura what do you thin about the idea?

**Comment by tmanninger** [comment]
@willyborankin do you mean, the Sink should create the Template, Datastream and ISM policy? 
If i am using the the current internal sink with an index, i also must create the index template with an alias manually.

**Comment by willyborankin** [comment]
> @willyborankin do you mean, the Sink should create the Template, Datastream and ISM policy? If i am using the the current internal sink with an index, i also must create the index template with an alias manually.

@tmanninger yes I think so, the main reason is that if you already has an audit index in your cluster and the cluster was migrated on the new version ... you fix will start to create a datastream ... which is not backward compatible.

**Comment by tmanninger** [comment]
> > @willyborankin do you mean, the Sink should create the Template, Datastream and ISM policy? If i am using the the current internal sink with an index, i also must create the index template with an alias manually.
> 
> @tmanninger yes I think so, the main reason is that if you already has an audit index in your cluster and the cluster was migrated on the new version ... you fix will start to create a datastream ... which is not backward compatible.

Really?
`irb.setOpType(DocWriteRequest.OpType.CREATE);`
Without an datastream template, the fix will not create a datastream. I tested it in my test environment.

**Comment by DarshitChanpura** [comment]
@tmanninger to ensure backwards compatibility for your fix: 
> Really?
irb.setOpType(DocWriteRequest.OpType.CREATE);
Without an datastream template, the fix will not create a datastream. I tested it in my test environment.

Can we add backwards compatibility tests as well?

**Comment by tmanninger** [comment]
I will add some test cases.

**Comment by tmanninger** [comment]
I added a test for datastream and for an index.
For both, i generate some audit logs and check and check the count of documents in the audit log index/datastream.

Thank's ok?

File: src/test/java/org/opensearch/security/auditlog/integration/BasicAuditlogTest.java
Comment (peternied) [inline src/test/java/org/opensearch/security/auditlog/integration/BasicAuditlogTest.java:588]: Getting the broad stats is useful, but we want to be sure the correct auto log entries are being generated, lets query the datastream's contents and be sure we have the expected audit log entries.

If the specific entries are missing the tests should fail, example testing depth [1]

- [1] https://github.com/opensearch-project/security/blob/main/src/test/java/org/opensearch/security/auditlog/compliance/ComplianceAuditlogTest.java#L309

File: src/test/java/org/opensearch/security/auditlog/integration/BasicAuditlogTest.java
Comment (peternied) [inline src/test/java/org/opensearch/security/auditlog/integration/BasicAuditlogTest.java:555]: I think we are missing a testing datatream's `_rollover` action - can you add validation for this kind of case where audit entries are created before/during/after and we are sure they have arrived intact?

File: src/test/java/org/opensearch/security/auditlog/integration/BasicAuditlogTest.java
Comment (peternied) [inline src/test/java/org/opensearch/security/auditlog/integration/BasicAuditlogTest.java:589]: Take this feedback or leave it: `assertThat` provides action error output on failure, especially when authoring tests: 
```suggestion
        var min = 3; var max = 5; // from 
        // ...
        assertThat(Integer.valueOf(res3.getTextFromJsonBody("/_all/primaries/docs/count")), allOf(greaterThan(min),lessThan(max)));
```

This would result in a message like 

```
//  java.lang.AssertionError:
//  Expected: (a value greater than <3> and a value less than <5>)
//       got: <1>
```


- [1] https://stackoverflow.com/a/7956297

**Review by peternied — COMMENTED** [review]
> I think the right solution would be to extend existing InternalOpenSearchSink or create a new Sync call it let's say internal_opensearch_datastream or something else. @peternied, @DarshitChanpura what do you thin about the idea?

@willyborankin This is a great idea - I think we should do this.

> Add a backward combability test.

I don't think a BWC test is needed in this scenario so long as we are adding a new audit log type.

@tmanninger Let us know if you have questions, while the proposed fix could work.  I'd rather there is consistency.  Having to correctly configure the data stream before its used seems like a way to have a broken audit log and not be aware of it, whereas having a concrete definition ensures that the data stream is setup correctly without user intervention

**Comment by tmanninger** [comment]
I can create a new Sink named internal_opensearch_datastream

- Create index template, if not exists
- Create datastream
- Write audit logs to this datastream

is this way ok?

**Comment by peternied** [comment]
@tmanninger Please look into how the other audit log storage systems [1] work and consider what should be have sensible defaults that is configurable vs required.

As far as sink names go I think `internal_opensearch` and `external_opensearch` would look silly with `external_opensearch_data_stream` side-by-side.  You might need to be creative with how the class is instantiation or create 'inner classes' one for index and one for data_stream

As you mentioned `plugins.security.audit.config.datastream` would effectively communicate the name, and I would recommend that `plugins.security.audit.config.index` is mutually exclusive.  I think the template values should be configurable via a similar mechanism for the index shard/replica count of an index

Looks forward to those changes - thank you.

- [1] https://opensearch.org/docs/latest/security/audit-logs/storage-types/

**Comment by tmanninger** [comment]
How can i create a datastream template?

When i am using this example:
https://github.com/opensearch-project/opensearch-java/blob/main/samples/src/main/java/org/opensearch/client/samples/DataStreamBasics.java.

`error: package org.opensearch.client.opensearch.indices does not exist`

I can use `org.opensearch.client.indices.PutIndexTemplateRequest`, but this class does not support DataStream config.

**Comment by DarshitChanpura** [comment]
@tmanninger Are you still working on this?

**Comment by tmanninger** [comment]
> How can i create a datastream template?
> 
> When i am using this example: https://github.com/opensearch-project/opensearch-java/blob/main/samples/src/main/java/org/opensearch/client/samples/DataStreamBasics.java.
> 
> `error: package org.opensearch.client.opensearch.indices does not exist`
> 
> I can use `org.opensearch.client.indices.PutIndexTemplateRequest`, but this class does not support DataStream config.

@DarshitChanpura see my post above... I find no option to create a IndexTemplate with datastream config...

**Comment by DarshitChanpura** [comment]
@tmanninger Did you try importing java-client? If yes, Can you upload your code set-up here?

**Review by DarshitChanpura — COMMENTED** [review]
@tmanninger Did you try importing java-client? If yes, Can you upload your code set-up here?

**Comment by tmanninger** [comment]
When i import the java client:

```error: package org.opensearch.client.opensearch.indices does not exist```

**Comment by DarshitChanpura** [comment]
@tmanninger Do you plan to work on this?

**Comment by tmanninger** [comment]
When nobody helps me resolving my issue, no.

Regards,
Thomas

**Comment by DarshitChanpura** [comment]
@tmanninger Can you add this line in build.gradle:
```
implementation('org.opensearch.client:opensearch-java:2.14.0')
```

and run :
```
./gradlew clean build -x test -x integrationTest -x spotbugsIntegrationTest
```

if needed add following lines where other resolutions are forced in build.gradle:
```
force "commons-logging:commons-logging:1.3.4"
force "org.opensearch.client:opensearch-rest-client:3.0.0-SNAPSHOT"
force "jakarta.json:jakarta.json-api:2.1.3"
```

This will resolve import issue.

**Comment by tmanninger** [comment]
@DarshitChanpura thanks for feedback, now i can import it.

I hope, the last question:

My sink provider (InternalOpenSearchDataStreamSink) get passed the client as parameter. This legacy- client cannot create datastreams and datastreams template as shown in the example:
https://github.com/opensearch-project/opensearch-java/blob/main/samples/src/main/java/org/opensearch/client/samples/DataStreamBasics.java

How can i convert the client to a [RestHighLevelClient](https://javadoc.io/static/org.opensearch.client/opensearch-rest-high-level-client/2.1.0/org/opensearch/client/RestHighLevelClient.html) or a IndicesClient or what's the correct way?

Simple Code example:

InternalOpenSearchDataStreamSink.java:
```
import org.opensearch.client.IndicesClient;
import org.opensearch.client.indices.CreateDataStreamRequest;

public final class InternalOpenSearchDataStreamSink extends AuditLogSink {

    public InternalOpenSearchDataStreamSink(
        ...
        final Client clientProvider,
        ...
    ) {
      ...
      CreateDataStreamRequest createDataStreamRequest = new CreateDataStreamRequest(dataStreamName);
      // Now i need an IndicesClient object to call IndicesClient.createDataStream()
      // clientProvider does not have a indices() method.
    )
```

**Comment by cwperks** [comment]
@tmanninger Can you use `clientProvider.admin().indices().createDataStream(Request...)`?

**Comment by tmanninger** [comment]
@DarshitChanpur @peternied 
I commited a new version:

Config options:
```
plugins.security.audit.type: internal_opensearch_data_stream
plugins.security.audit.config.data_stream.name: opensearch-security-auditlog # Datastream name
plugins.security.audit.config.data_stream.template.manage: true # Should the template managed by the plugin?
plugins.security.audit.config.data_stream.template.name: opensearch-security-auditlog # Name of the template
plugins.security.audit.config.data_stream.template.number_of_shards: 1 # Number of shards
plugins.security.audit.config.data_stream.template.number_of_replicas: 0 # Number of repliaces
```

One note: I cannot create the template in the `InternalOpenSearchDataStreamSink` constructor, because at this step, the node is not up and ready.
I moved the creation of the template in the "datastream" funktion.

**Comment by peternied** [comment]
@tmanninger Thanks for circling back to this pull request.  I won't have time to re-review so I've dismissed my review, the other maintainers @cwperks & @DarshitChanpura should be able to help you out.

**Comment by tmanninger** [comment]
Tomorrow i will fix the spotline and checkstyle errors.

File: src/main/java/org/opensearch/security/auditlog/sink/InternalOpenSearchDataStreamSink.java
Comment (cwperks) [inline src/main/java/org/opensearch/security/auditlog/sink/InternalOpenSearchDataStreamSink.java:48]: There may be an opportunity to abstract the logic that is common to here and InternalOpenSearchSink into an Abstract class that both classes can inherit. These classes are pretty small, so it won't reduce the code too much but I thought I'd leave a comment  none-the-less.

File: src/main/java/org/opensearch/security/auditlog/sink/InternalOpenSearchDataStreamSink.java
Comment (cwperks) [inline src/main/java/org/opensearch/security/auditlog/sink/InternalOpenSearchDataStreamSink.java:93]: Is there a Builder pattern somewhere that can help assemble this payload?

I haven't looked at it in detail, but something like this: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/internalClusterTest/java/org/opensearch/action/admin/indices/datastream/DataStreamTestCase.java#L81-L89

File: src/main/java/org/opensearch/security/auditlog/sink/InternalOpenSearchDataStreamSink.java
Comment (tmanninger) [reply src/main/java/org/opensearch/security/auditlog/sink/InternalOpenSearchDataStreamSink.java]: I created an abstracted logic with AbstractInternalOpenSearchSink and shared duplicate code in this class.

File: src/main/java/org/opensearch/security/auditlog/sink/InternalOpenSearchDataStreamSink.java
Comment (tmanninger) [reply src/main/java/org/opensearch/security/auditlog/sink/InternalOpenSearchDataStreamSink.java]: Yes, that's really ugly, but i found no other way. Thanks for the link, i fixed it.

**Comment by tmanninger** [comment]
I rebased git in an incorrect way :-/ (and github was auto-closed this MR) ...
I readded all my changes in one commit

**Comment by tmanninger** [comment]
Added MR for documentation: https://github.com/opensearch-project/documentation-website/pull/8356

**Comment by tmanninger** [comment]
I fixed the spotless errors.

**Comment by tmanninger** [comment]
Test
```
    Expected: (a value greater than <0> and a value less than <5>)
         but: a value less than <5> <5> was equal to <5>
        at __randomizedtesting.SeedInfo.seed([BB9BC69071BEFE5B:AF3E69E787FB9BB0]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
        at org.opensearch.security.auditlog.sink.InternalOpensearchDataStreamSinkTest.testTemplate(InternalOpensearchDataStreamSinkTest.java:108)
        at org.opensearch.security.auditlog.sink.InternalOpensearchDataStreamSinkTest.testDefaultSettings(InternalOpensearchDataStreamSinkTest.java:123)
 ```

Sometimes, there is one more audit log in the new datastream backend, maybe the rollover is faster than writing all audit logs generated before the rollover.
I increased the number of maximal audit docs.

@willyborankin pls rerun (hope the last time).

File: src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java
Comment (cwperks) [inline src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java:1623]: Curious, why a default value of 3 for this setting?

**Review by cwperks — DISMISSED** [review]
> Added MR for documentation: [opensearch-project/documentation-website#8356](https://github.com/opensearch-project/documentation-website/pull/8356)

Thank you for creating a documentation PR for this change!

The changes look good to me, I just had one remaining question on the default value for number of shards.

**Comment by tmanninger** [comment]
> The changes look good to me, I just had one remaining question on the default value for number of shards.

ss4o_traces_template and ss4o_metrics_template are using 3 shards as default value on all my opensearch clusters... So i copied the value.

File: src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java
Comment (tmanninger) [reply src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java:1623]: ss4o_traces_template and ss4o_metrics_template are using 3 shards as default value in opensearch... So i copied the value.

Comment (cwperks) [reply src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java:1623]: wdyt about setting this to `1` by default? 

There was an issue recently filed about the security audit log index coming up as yellow in single node clusters: https://github.com/opensearch-project/security/issues/4739

By setting primary shard count 1 and replica shard count 0 by default (and allowing customization), it will come up as green in single node cluster.

Comment (tmanninger) [reply src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java]: I changed the default shard value to 1

**Review by cwperks — APPROVED** [review]

**Review by willyborankin — APPROVED** [review]

**Review by DarshitChanpura — APPROVED** [review]

**Pull request merged**
